phac-nml/irida-next
# Data Export - Create Job

## What does this PR do and why?
This PR adds the Data Export `CreateJob`. Because exports have the potential to be very large files, we first initiate the export creation in the `CreateService`, which then calls the `CreateJob` to create and attach the actual export file in the background.

Note: Email notification will be handled in a [separate PR](https://github.com/phac-nml/irida-next/pull/464)
Note: Analysis exports will be added when analysis storage is implemented. 

## Screenshots or screen recordings
_Screenshots are required for UI changes, and strongly recommended for all other pull requests._

## How to set up and validate locally

- Open `rails c`
- Reseed your database if needed `bin/rails db:drop/create/migrate/seed`

- Set up your params:
```
user = User.find_by(email: 'user1@email.com')
params_1 = {'export_type' => 'sample', 'export_parameters' => {'ids' => [Sample.find_by(name: 'Streptococcus agalactiae/Outbreak 2022 Sample 10').id, Sample.find_by(name: 'Streptococcus agalactiae/Outbreak 2022 Sample 9').id, Sample.find_by(name: 'Neisseria gonorrhoeae/Outbreak 2022 Sample 10').id]}, 'email_notification' => true, 'name' => 'test export'}
```
- Navigate to `Streptococcus agalactiae/Outbreak 2022 Sample 10` sample page, and upload:
  - A set of paired end attachments (can be found in `test/fixtures/files`). An example of files would be `08-5578-small_S1_L001_R1_001.fastq` and `08-5578-small_S1_L001_R2_001.fastq`
  - A unknown file format (not a `fasta` or `fastq` file)

- Create an export with: `DataExports::CreateService.new(user, params_1).execute`
- Set: `data_export = user.data_exports.last`
- After 30 seconds, ensure `data_export.status = ready` and `data_export.expires_at` is 3 business days from the current date

Download the zip file
- To download the zip file, in the console run: `Rails.application.routes.url_helpers.rails_blob_url(data_export.file)`
- Copy and paste the above output into a browser. This should download the zip.
- Check Zip File:
  - Your file directories should be the following:
```
|---- manifest.json
|---- S. agalactiae outbreak 2022 project PUID
|                          |--------- Sample 10 PUID
|                          |                   | ---- test_file_1.fastq PUID
|                          |                   |                      |------- test_file_1.fastq
|                          |                   |   
|                          |                   | ---- unknown file PUID
|                          |                   |                      |------- unknown file.format
|                          |                   |   
|                          |                   | ---- paired files PUID
|                          |                                          |------- paired file 1
|                          |                                          |------- paired file 2
|                          |  
|                          |--------- Sample 9 PUID
|                                              | ---- test_file_1.fastq PUID
|                                                                     |------- test_file_1.fastq
|---- N. gonorrhoeae/Outbreak 2022 project PUID
                           |--------- Sample 10 PUID
                                               | ---- test_file_1.fastq PUID
                                                                     |------- test_file_1.fastq
```

  - Check the `manifest.json` is the following:
```
{
  "type": "Samples Export",
  "date": "Date.current",
  "children": [
    {
      "name": "PUID S. agalactiae Outbreak 2022",
      "type": "folder",
      "irida-next-type": "project",
      "irida-next-name": "Outbreak 2022",
      "children": [
        {
          "name": "PUID Streptococcus agalactiae/Outbreak 2022 Sample 10",
          "type": "folder",
          "irida-next-type": "sample",
          "irida-next-name": "Streptococcus agalactiae/Outbreak 2022 Sample 10",
          "children": [
            {
              "name": "PUID Paired End files",
              "type": "folder",
              "irida-next-type": "attachment",
              "children": [
                {
                  "name": "08-5578-small_S1_L001_R1_001.fastq",
                  "type": "file",
                  "metadata": {
                    "format": "fastq",
                    "direction": "forward",
                    "type": "illumina_pe"
                  }
                },
                {
                  "name": "08-5578-small_S1_L001_R2_001.fastq",
                  "type": "file",
                  "metadata": {
                    "format": "fastq",
                    "direction": "reverse",
                    "type": "illumina_pe"
                  }
                }
              ]
            },
            {
              "name": "PUID unknown file",
              "type": "folder",
              "irida-next-type": "attachment",
              "children": [
                {
                  "name": "Screenshot from 2024-03-11 11-42-46.png",
                  "type": "file",
                  "metadata": {
                    "format": "unknown"
                  }
                }
              ]
            },
            {
              "name": "PUID test_file_1.fastq",
              "type": "folder",
              "irida-next-type": "attachment",
              "children": [
                {
                  "name": "test_file_1.fastq",
                  "type": "file",
                  "metadata": {
                    "format": "fastq"
                  }
                }
              ]
            }
          ]
        },
        {
          "name": "PUID Streptococcus agalactiae/Outbreak 2022 Sample 9",
          "type": "folder",
          "irida-next-type": "sample",
          "irida-next-name": "Streptococcus agalactiae/Outbreak 2022 Sample 9",
          "children": [
            {
              "name": "PUID test_file_1.fastq",
              "type": "folder",
              "irida-next-type": "attachment",
              "children": [
                {
                  "name": "test_file_1.fastq",
                  "type": "file",
                  "metadata": {
                    "format": "fastq"
                  }
                }
              ]
            }
          ]
        }
      ]
    },
    {
      "name": "PUID Neisseria gonorrhoeae Outbreak 2022 Project",
      "type": "folder",
      "irida-next-type": "project",
      "irida-next-name": "Outbreak 2022",
      "children": [
        {
          "name": "PUID Neisseria gonorrhoeae/Outbreak 2022 Sample 10",
          "type": "folder",
          "irida-next-type": "sample",
          "irida-next-name": "Neisseria gonorrhoeae/Outbreak 2022 Sample 10",
          "children": [
            {
              "name": "PUID test_file_1.fastq",
              "type": "folder",
              "irida-next-type": "attachment",
              "children": [
                {
                  "name": "test_file_1.fastq",
                  "type": "file",
                  "metadata": {
                    "format": "fastq"
                  }
                }
              ]
            }
          ]
        }
      ]
```
## PR acceptance checklist
This checklist encourages us to confirm any changes have been analyzed to reduce risks in quality, performance, reliability, security, and maintainability.
